File: 10_data_management/fixer/testme.R
#' @name testme.R
#' @title Smoke test Ollama Cloud chat from the fixer folder (no tools)
#' @author Prof. Tim Fraser
#' @description
#' One POST to **`/api/chat`** with a tiny user message — verifies **`OLLAMA_API_KEY`**,
#' **`OLLAMA_HOST`**, and **`OLLAMA_MODEL`** without tool calling or **`num_predict`**
#' (helps isolate HTTP 400s from the full **`fixer_csv.R`** loop).

# 0. SETUP ###################################

resolve_fixer_root = function() {
  r = Sys.getenv("FIXER_ROOT", unset = "")
  if (nzchar(r) && dir.exists(r)) {
    return(normalizePath(r, winslash = "/", mustWork = TRUE))
  }
  wd = normalizePath(getwd(), winslash = "/", mustWork = TRUE)
  if (file.exists(file.path(wd, "functions.R"))) {
    return(wd)
  }
  cand = file.path(wd, "10_data_management", "fixer")
  if (file.exists(file.path(cand, "functions.R"))) {
    return(normalizePath(cand, winslash = "/", mustWork = TRUE))
  }
  stop("Run from fixer/, dsai repo root, or set FIXER_ROOT.")
}

FIXER_ROOT = resolve_fixer_root()
env_path = file.path(FIXER_ROOT, ".env")
if (file.exists(env_path)) {
  readRenviron(env_path)
}

OLLAMA_HOST = trimws(Sys.getenv("OLLAMA_HOST", unset = "https://ollama.com"))
OLLAMA_API_KEY = trimws(Sys.getenv("OLLAMA_API_KEY", unset = ""))
OLLAMA_MODEL = trimws(Sys.getenv("OLLAMA_MODEL", unset = ""))
if (!nzchar(OLLAMA_MODEL)) {
  OLLAMA_MODEL = "gpt-oss:120b"
}

if (!nzchar(OLLAMA_API_KEY)) {
  stop("Set OLLAMA_API_KEY in fixer/.env (copy from .env.example).")
}

source(file.path(FIXER_ROOT, "functions.R"), local = FALSE)

# 1. MINIMAL CHAT (test query) ###################################

# Deliberately tiny: model should answer in a few tokens; no tools, no options.
messages = list(
  list(
    role = "user",
    content = "Reply with exactly one word: pong"
  )
)

cat("Test query: POST ", OLLAMA_HOST, "/api/chat\n", sep = "")
cat("Model: ", OLLAMA_MODEL, "\n\n", sep = "")

out = tryCatch(
  ollama_chat_once(
    OLLAMA_HOST,
    OLLAMA_API_KEY,
    OLLAMA_MODEL,
    messages,
    tools = NULL,
    format = NULL,
    max_output_tokens = NULL
  ),
  error = function(e) {
    lr = tryCatch(httr2::last_response(), error = function(e2) NULL)
    if (!is.null(lr)) {
      cat("HTTP status:", httr2::resp_status(lr), "\n")
      b = tryCatch(httr2::resp_body_string(lr), error = function(e3) "(no body)")
      cat("Response body:\n", b, "\n", sep = "")
    }
    stop(e)
  }
)

cat("Assistant content:\n", out$content, "\n\n", sep = "")
cat("Smoke test OK.\n")

File: 10_data_management/fixer/testme.py
# testme.py
# Smoke test Ollama Cloud chat from the fixer folder (no tools)
# Tim Fraser

from __future__ import annotations

import os
import sys

import httpx
from dotenv import load_dotenv

from pathlib import Path

_FIXER_ROOT = Path(__file__).resolve().parent
sys.path.insert(0, str(_FIXER_ROOT))


def resolve_fixer_root() -> Path:
    r = os.environ.get("FIXER_ROOT", "").strip()
    if r and Path(r).is_dir():
        return Path(r).resolve()
    wd = Path.cwd().resolve()
    if (wd / "functions.py").is_file() or (wd / "functions.R").is_file():
        return wd
    cand = wd / "10_data_management" / "fixer"
    if (cand / "functions.py").is_file():
        return cand.resolve()
    raise SystemExit("Run from fixer/, dsai repo root, or set FIXER_ROOT.")


from functions import ollama_chat_once

FIXER_ROOT = resolve_fixer_root()
env_path = FIXER_ROOT / ".env"
if env_path.is_file():
    load_dotenv(env_path)

OLLAMA_HOST = os.environ.get("OLLAMA_HOST", "https://ollama.com").strip()
OLLAMA_API_KEY = os.environ.get("OLLAMA_API_KEY", "").strip()
OLLAMA_MODEL = os.environ.get("OLLAMA_MODEL", "").strip()
if not OLLAMA_MODEL:
    OLLAMA_MODEL = "gpt-oss:120b"

if not OLLAMA_API_KEY:
    raise SystemExit("Set OLLAMA_API_KEY in fixer/.env (copy from .env.example).")

messages = [
    {"role": "user", "content": "Reply with exactly one word: pong"},
]

print(f"Test query: POST {OLLAMA_HOST}/api/chat")
print(f"Model: {OLLAMA_MODEL}\n")

try:
    out = ollama_chat_once(
        OLLAMA_HOST,
        OLLAMA_API_KEY,
        OLLAMA_MODEL,
        messages,
        tools=None,
        format=None,
        max_output_tokens=None,
    )
except Exception as e:
    try:
        # httpx stores last response on exception in some cases
        if isinstance(e, httpx.HTTPStatusError) and e.response is not None:
            print("HTTP status:", e.response.status_code)
            print("Response body:\n", e.response.text)
    except Exception:
        pass
    raise SystemExit(str(e)) from e

print("Assistant content:\n", out.get("content", ""), "\n", sep="")
print("Smoke test OK.")
